File: 135-candy/candy.py
class Solution:
    def candy(self, ratings: List[int]) -> int:
        # loop 3 times

        n = len(ratings)
        candies = [1] * n # start each child with 1 candy
        
        # first loop compare with LH neighbour (1 -> n-1) 
        for i in range(1, n):
            if ratings[i] > ratings[i-1]:
                candies[i] = candies[i-1] + 1
        
        # 2nd loop compare with RH neighbour (n-2 -> 0)
        for i in range(n-2, -1, -1):
            if ratings[i] > ratings[i+1] and candies[i] <= candies[i+1]:
                candies[i] = candies[i+1] + 1

        # last time to calculate number of candies
        return sum(candies)

File: 135-candy/candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        vector<int> candies(n, 1); // each child start with 1

        // compare to left neighbour
        for (int i = 1; i < n; ++i) {
            if (ratings[i] > ratings[i-1]) {
                candies[i] = candies[i-1] + 1;
            }
        }

        // compare to right neighbour
        for (int i = n-2; i >= 0; --i) {
            if (ratings[i] > ratings[i+1]) {
                // if have a higher rating than their RH neighbour, they will get 1 more candy than RH neighbour unless they already have more candies than their RH
                candies[i] = max(candies[i], candies[i+1] + 1);
            }
        }

        int res = 0;
        for (int c : candies) {
            res += c;
        }

        return res;
    }
};